spring-cloud-task-core/src/main/java/org/springframework/cloud/task/configuration/SingleInstanceTaskListener.java

```diff
@@ -0,0 +1,154 @@
+
+	private TaskNameResolver taskNameResolver;
+
+	private ApplicationEventPublisher applicationEventPublisher;
+
+	private boolean lockReady;
+
+	private boolean lockFailed;
+
+	private DataSource dataSource;
+
+	private TaskProperties taskProperties;
+
+	public SingleInstanceTaskListener(LockRegistry lockRegistry,
+			TaskNameResolver taskNameResolver,
+			TaskProperties taskProperties,
+			ApplicationEventPublisher applicationEventPublisher) {
+		this.lockRegistry = lockRegistry;
+		this.taskNameResolver = taskNameResolver;
+		this.taskProperties = taskProperties;
+		this.lockRegistryLeaderInitiator = new LockRegistryLeaderInitiator(this.lockRegistry);
+		this.applicationEventPublisher = applicationEventPublisher;
+	}
+
+	public SingleInstanceTaskListener(DataSource dataSource,
+			TaskNameResolver taskNameResolver,
+			TaskProperties taskProperties,
+			ApplicationEventPublisher applicationEventPublisher) {
+		this.taskNameResolver = taskNameResolver;
+		this.applicationEventPublisher = applicationEventPublisher;
+		this.dataSource = dataSource;
+		this.taskProperties = taskProperties;
+	}
+
+	@BeforeTask
+	public void lockTask(TaskExecution taskExecution) {
+		if(this.lockRegistry == null ) {
+			this.lockRegistry = getDefaultLockRegistry(taskExecution.getExecutionId());
+		}
+		this.lockRegistryLeaderInitiator = new LockRegistryLeaderInitiator(
+				this.lockRegistry,
+				new DefaultCandidate(String.valueOf(taskExecution.getExecutionId()),
+						taskNameResolver.getTaskName()));
+		this.lockRegistryLeaderInitiator.setApplicationEventPublisher(this.applicationEventPublisher);
+		this.lockRegistryLeaderInitiator.setPublishFailedEvents(true);
+		this.lockRegistryLeaderInitiator.start();
+		while (!this.lockReady) {
+			try {
+				Thread.sleep(this.taskProperties.getSingleInstanceLockCheckInterval());
+			}
+			catch (InterruptedException ex) {
+				logger.warn("Thread Sleep Failed", ex);
+			}
+			if (this.lockFailed) {
+				String errorMessage = String.format(
+						"Task with name \"%s\" is already running.",
+						this.taskNameResolver.getTaskName());
+				try {
+					this.lockRegistryLeaderInitiator.destroy();
+				}
+				catch (Exception exception) {
+					throw new TaskExecutionException("Failed to destroy lock.", exception);
+				}
+				throw new TaskExecutionException(errorMessage);
+			}
+		}
+	}
+
+	@AfterTask
+	public void unlockTaskOnEnd(TaskExecution taskExecution) throws Exception {
+		this.lockRegistryLeaderInitiator.destroy();
+	}
+
+	@FailedTask
+	public void unlockTaskOnError(TaskExecution taskExecution) throws Exception {
+		this.lockRegistryLeaderInitiator.destroy();
+	}
+
+	@Override
+	public void onApplicationEvent(ApplicationEvent applicationEvent) {
+		if (applicationEvent instanceof OnGrantedEvent) {
+			this.lockReady = true;
+		}
+		else if (applicationEvent instanceof OnFailedToAcquireMutexEvent) {
+			this.lockFailed = true;
+		}
+	}
+
+	private LockRegistry getDefaultLockRegistry( long executionId) {
```

Comment:
Is there a reason for this being lazily created and not just done in the constructor that doesn't set a registry?

Reply:
We use the executionId as a unique id for the lock registry.  This is not available at constructor time.

Reply:
I missed that. Thanks.